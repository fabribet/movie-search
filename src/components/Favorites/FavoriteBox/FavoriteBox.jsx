import React from 'react'
import PropTypes from 'prop-types'
import styles from './styles.module.scss'
import Truncate from 'react-truncate'

import moviePoster from './imgs/movie-poster.png'

/**
 * FavoriteBox - React component.
 * Renders a small movie box .
 *
 * Properties
 * - imgUrl {String}       - The base url to get movie posters. {Required}
 * - removeFavorite {func} - Allows to remove this favorite movie from the list. {Required}
 * - movie {Object}        - The movie information
 *    - id
 *    - title
 *    - poster_path
 */
function FavoriteBox(props) {

  /**
   * Handles the remove button. Calls the removing method.
   * @param {window.Event} e 
   */
  const onFavoriteClick = (e) => {
    e.cancelBubble = true;
    if (e.stopPropagation) e.stopPropagation();
    props.removeFavorite(props.movie.id)
  }

  /**
   * Handles the box click. Opens a new window with searching for the movie trailer on YouTube.
   * @param {window.Event} e 
   */
  const onMovieClick = (e) => {
    const windowFeatures = 'menubar=no,location=no,resizable=no,scrollbars=yes,status=yes,height=400,width=400'
    window.open(
      `https://www.youtube.com/results?search_query=${props.movie.title}+movie+trailer`,
      `${props.movie.title} trailer`,
      windowFeatures
    )
  }

  return (
    <div className={styles.box} onClick={onMovieClick}>
      <div className={styles.img}  style={{
        backgroundImage: props.imgUrl && props.movie.poster_path 
          ? `url(${props.imgUrl}w154${props.movie.poster_path})`
          : `url(${moviePoster})`
        }}
      />
      <div className={styles.movieInfo}>
        <div className={styles.titleWrapper}>
          <h4 className={styles.title}>
            <Truncate lines={2} ellipsis="...">
              {props.movie.title}
            </Truncate>
          </h4>
          <div id={'fb-removebtn'} className={styles.removeBtn} title={'Remove favorite'} onClick={onFavoriteClick}/>
        </div>
      </div>
    </div>
  )
}

FavoriteBox.propTypes = {
    imgUrl: PropTypes.string,
    removeFavorite: PropTypes.func.isRequired,
    movie: PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      poster_path: PropTypes.string.isRequired
    }).isRequired
}

export default FavoriteBox