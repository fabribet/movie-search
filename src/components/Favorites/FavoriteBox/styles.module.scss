.box {
  display: flex;
  flex-shrink: 0;
  font-size: 14px;
  margin-right: 10px;
  height: 51.75px;
  width: 150px;
  cursor: pointer;
  -webkit-box-shadow: 2px 2px 5px 0px rgba(0,0,0,0.75);
  -moz-box-shadow: 2px 2px 5px 0px rgba(0,0,0,0.75);
  box-shadow: 2px 2px 5px 0px rgba(0,0,0,0.75);

  .img {
    width: 34.5px;
    height: 51.75px;
    flex-shrink: 0;
    background-size: contain;
    background-repeat: no-repeat;
  }

  .movieInfo {
    padding:  5px  0;
    flex-grow: 1;
    
    .titleWrapper {
      display: flex;

      .title {
        flex-grow: 1;
        font-size: 10px;
      }

      .removeBtn {
        flex-shrink: 0;
        width: 15px;
        height: 15px;
        margin-left: 5px;
        background-size: contain;
        background-repeat: no-repeat;
        flex-shrink: 0;
        background-image: url('./imgs/times-solid.svg');
      }
    }
  }

}

