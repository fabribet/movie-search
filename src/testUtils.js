import configureStore from 'redux-mock-store'

// Mock the canvas usage for external library usage on tests.
export function mockCanvas () {
  const createElement = document.createElement.bind(document)
  document.createElement = tagName => {
    const element = createElement(tagName)
    if (tagName === 'canvas') {
      element.getContext = () => ({})
    }
    return element
  }
}

/**
 * Create Mock store for testing
 */
const mockStore = configureStore([])
export function createMockStore (initialState = {
  Favorites: {
    favorites: {}
  },
  TMDB: {
    apiConfig: {
      loading: false,
      error: null,
      data: {
        images: {
          base_url: ''
        }
      }
    }
  }
}) {
  return mockStore(initialState)
}

export const FAVORITES_DATA = {
  1: {
    id: 1,
    title: 'Movie 1',
    poster_path: '/asdasdas.jpg'
  },

  2: {
    id: 2,
    title: 'Movie 2',
    poster_path: '/asdasdas2.jpg'
  }
}
