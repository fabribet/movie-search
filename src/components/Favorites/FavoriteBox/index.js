import { connect } from 'react-redux'
import { actions } from '../../../actions/Favorites'
import FavoriteBox from './FavoriteBox'

const mapStateToProps = state => ({
  // If the API configuration data has been obtained the Images base url is passed.
  imgUrl: state.TMDB.apiConfig.data ? state.TMDB.apiConfig.data.images.base_url : null
})

const mapDispatchToProps = dispatch => ({
  removeFavorite (movieId) {
    dispatch(actions.RemoveFavorite(movieId))
  }
})

export default connect(mapStateToProps, mapDispatchToProps)(FavoriteBox)
