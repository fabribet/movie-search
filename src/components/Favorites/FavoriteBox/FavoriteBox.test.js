import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
// import { render } from '@testing-library/react'
import FavoriteBox from './FavoriteBox.jsx'
import { mockCanvas } from './../../../testUtils'

mockCanvas()

const SpiedRemoveFavorite = jest.fn()
const favoriteBox = <FavoriteBox removeFavorite={SpiedRemoveFavorite} movie={{ id: 1, title: 'fake movie', poster_path: '/bla.png' }} />

test('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(
    favoriteBox,
    div
  )
})

it('Renders favorite box correctly', () => {
  const tree = renderer
    .create(favoriteBox)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Remove Favorite is called when the remove btn is clicked', () => {
  // Render a checkbox with label in the document
  const shallowFB = shallow(favoriteBox)
  const closeBtn = shallowFB.find('div#fb-removebtn')
  const eventObj = {}
  closeBtn.simulate('click', eventObj)

  expect(SpiedRemoveFavorite).toHaveBeenCalled()
})
