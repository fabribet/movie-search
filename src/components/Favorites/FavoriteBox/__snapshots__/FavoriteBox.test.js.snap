// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders favorite box correctly 1`] = `
<div
  className="box"
  onClick={[Function]}
>
  <div
    className="img"
    style={
      Object {
        "backgroundImage": "url(movie-poster.png)",
      }
    }
  />
  <div
    className="movieInfo"
  >
    <div
      className="titleWrapper"
    >
      <h4
        className="title"
      >
        <span
          width={0}
        >
          <span />
          <span>
            fake movie
          </span>
          <span
            style={
              Object {
                "left": 0,
                "position": "fixed",
                "top": 0,
                "visibility": "hidden",
              }
            }
          >
            ...
          </span>
        </span>
      </h4>
      <div
        className="removeBtn"
        id="fb-removebtn"
        onClick={[Function]}
        title="Remove favorite"
      />
    </div>
  </div>
</div>
`;
